### PR TITLE
JIT: Fix LowerHWIntrinsicGetElement for xarch

### LINKED ISSUE #89565: JIT: LowerHWIntrinsicGetElement can reorder trees unsafely
The following example throws `ArgumentOutOfRangeException` when `Vector256` is hardware accelerated, but `NullReferenceException` when it isn't (`DOTNET_EnableAVX=0`).

```csharp
public class Program
{
    public static unsafe void Main()
    {
        Foo(null);
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static unsafe float Foo(Vector256<float>* v)
    {
        return (*v)[8];
    }
}
```

Context: https://github.com/dotnet/runtime/pull/83005#issuecomment-1653663453

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Ensure we don't reorder evaluation and change exceptional behavior.

Closes #89565.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@jakobbotsch PTAL
cc @dotnet/jit-contrib

File: src/coreclr/jit/lowerxarch.cpp [hunk 1]
```diff
@@ -5074,8 +5074,43 @@ GenTree* Lowering::LowerHWIntrinsicGetElement(GenTreeHWIntrinsic* node)
...
+                BlockRange().TryGetUse(addr, &addrUse);
+                unsigned addrLcl = addrUse.ReplaceWithLclVar(comp);
+                addr             = indir->Addr();
+            }
+
+            // If the indir can fault, do a null check.
+            //
+            if (indir->OperMayThrow(comp))
+            {
+                GenTree* addrClone = comp->gtClone(addr);
```
Comment (jakobbotsch) [inline src/coreclr/jit/lowerxarch.cpp:5102]:
```suggestion
                GenTree* addrClone = comp->gtCloneExpr(addr);
```

File: src/coreclr/jit/lowerxarch.cpp [hunk 2]
```diff
@@ -5074,8 +5074,43 @@ GenTree* Lowering::LowerHWIntrinsicGetElement(GenTreeHWIntrinsic* node)
...
+        // If we can't move the indir, force evaluation of its side effects.
+        //
+        if (!canMoveTheIndirLater)
+        {
+            // Force evaluation of the address, if it is complex
+            //
+            if (!(addr->IsInvariant() || addr->OperIsLocal()))
+            {
+                LIR::Use addrUse;
+                BlockRange().TryGetUse(addr, &addrUse);
```
Comment (jakobbotsch) [inline src/coreclr/jit/lowerxarch.cpp:5093]:
```suggestion
                LIR::Use addrUse(BlockRange(), &indir->Addr(), indir);
```

File: src/coreclr/jit/lowerxarch.cpp [hunk 3]
```diff
@@ -5074,8 +5074,43 @@ GenTree* Lowering::LowerHWIntrinsicGetElement(GenTreeHWIntrinsic* node)
...
+                unsigned addrLcl = addrUse.ReplaceWithLclVar(comp);
```
Comment (jakobbotsch) [inline src/coreclr/jit/lowerxarch.cpp:5094]:
```suggestion
                addrUse.ReplaceWithLclVar(comp);
```

**Review by jakobbotsch — APPROVED** [review]
LGTM otherwise. Were the diffs bad if we just skipped the optimization when we cannot move the indir?

File: src/tests/JIT/Regression/JitBlue/Runtime_89565/Runtime_89595.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (EgorBo) [inline src/tests/JIT/Regression/JitBlue/Runtime_89565/Runtime_89595.cs:1]: nit: wrong issue id for file name - 89595 vs 89565

Comment (AndyAyersMS) [reply src/tests/JIT/Regression/JitBlue/Runtime_89565/Runtime_89595.cs:1]: Ah, thanks.

**Comment by AndyAyersMS** [comment]
> LGTM otherwise. Were the diffs bad if we just skipped the optimization when we cannot move the indir?

I don't recall -- let me look.

**Comment by AndyAyersMS** [comment]
From SPMI the diffs are somewhat bigger, eg in ASP.NET:

```
With Fix

[08:46:54] Top method regressions (bytes):
[08:46:54]            6 (13.33% of base) : 66717.dasm - System.Numerics.Vector`1[int]:get_Item(int):int:this (Instrumented Tier0)
[08:46:54]            6 (13.33% of base) : 10948.dasm - System.Numerics.Vector`1[int]:get_Item(int):int:this (Tier0)
[08:46:54]            6 (13.33% of base) : 20741.dasm - System.Numerics.Vector`1[int]:get_Item(int):int:this (Tier0)
[08:46:54]            6 (13.33% of base) : 64929.dasm - System.Runtime.Intrinsics.Vector128`1[int]:get_Item(int):int:this (Tier0)
[08:46:54]            6 (13.04% of base) : 58073.dasm - System.Runtime.Intrinsics.Vector128`1[long]:get_Item(int):long:this (Tier0)
[08:46:54]            6 (13.04% of base) : 36918.dasm - System.Runtime.Intrinsics.Vector128`1[long]:get_Item(int):long:this (Tier0)
[08:46:54]            6 (13.04% of base) : 64402.dasm - System.Runtime.Intrinsics.Vector256`1[long]:get_Item(int):long:this (Tier0)
[08:46:54]            2 ( 8.70% of base) : 152279.dasm - System.Numerics.Vector`1[int]:get_Item(int):int:this (Tier1)

With Avoid

[11:41:47] Top method regressions (bytes):
[11:41:47]           14 (60.87% of base) : 152279.dasm - System.Numerics.Vector`1[int]:get_Item(int):int:this (Tier1)
[11:41:47]           13 (28.89% of base) : 66717.dasm - System.Numerics.Vector`1[int]:get_Item(int):int:this (Instrumented Tier0)
[11:41:47]           13 (28.89% of base) : 20741.dasm - System.Numerics.Vector`1[int]:get_Item(int):int:this (Tier0)
[11:41:47]           13 (28.89% of base) : 10948.dasm - System.Numerics.Vector`1[int]:get_Item(int):int:this (Tier0)
[11:41:47]           13 (28.26% of base) : 64402.dasm - System.Runtime.Intrinsics.Vector256`1[long]:get_Item(int):long:this (Tier0)
[11:41:47]           10 (22.22% of base) : 64929.dasm - System.Runtime.Intrinsics.Vector128`1[int]:get_Item(int):int:this (Tier0)
[11:41:47]           10 (21.74% of base) : 36918.dasm - System.Runtime.Intrinsics.Vector128`1[long]:get_Item(int):long:this (Tier0)
[11:41:47]           10 (21.74% of base) : 58073.dasm - System.Runtime.Intrinsics.Vector128`1[long]:get_Item(int):long:this (Tier0)
```

image

There are only a few diffs in non `get_Item` methods.

**Comment by AndyAyersMS** [comment]
NAOT failure looks to be related:
```
  ILC: D:\a\_work\1\s\src\coreclr\jit\lsrabuild.cpp:2601
  ILC: Assertion failed '!"Expected empty defList at end of block"' in 'System.Numerics.Matrix4x4:get_Item(int,int):float:this' during 'Linear scan register alloc' (IL size 14; hash 0x67a5e306; FullOpts)
  ```

**Comment by AndyAyersMS** [comment]
Ah, other failures too.

**Comment by tannergooding** [comment]
@AndyAyersMS, this seems to have inserted quite a few null checks for the `ref this` parameter on structs:
```diff
        ; byrRegs +[rcx]
+       cmp      byte  ptr [rcx], cl
        cmp      edx, 4
        jae      SHORT G_M63163_IG04
        vmovsd   xmm0, qword ptr [rcx+8*rdx]
-						;; size=10 bbWeight=1 PerfScore 5.25
+						;; size=12 bbWeight=1 PerfScore 8.25
```

Is that expected? We normally don't see such null checks (see for example: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACXDKAVzAwYGUGBvGhoIYATCK2AAbGAxgAGANwCho8VJnlF1IQ3oixk6QHEYGAJoAKAJQMAvAD51mgL5A==)

**Comment by AndyAyersMS** [comment]
That's kind of the point. It adds a handful of nullchecks in places where the JIT doesn't or can't prove the vector address won't be null, since the dereference happens before the element access.

**Comment by tannergooding** [comment]
It looks to be generally inconsistent with how we handle it for spans and other cases: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4axAMwNcGKBzAYGshgG8aDKwwAmEDsGkMYdDdWu37jmM/JuPltYA2gCyMBgAFhA2AJLikgAUYZHRcWKSAPJifBBcuCwAchAxXJK8XOUA5gCUALqBVtqeDk4A4uEAGgnVDAC8AHwMXDAA7iZi2FwAPHYtMP0JsABmznTVQXS1/tYNTDqz3gztGACa3X2Dw2OyE9MH0gvLvuubbgC+NJo6ZAxCNBbuaxNbi4bBLHz3HxdWQAKgYYh6A3hcH6xy61W2Vl2wLyYIhXicZ1h8MRgzEKOOZwxNDeQA=

I understand the intent of doing null checks here and think that in a case like `p->GetElement(0)` it's goodness, but it also appears to be inconsistent with general handling of the implicit `this` parameter for structs and so it isn't clear to me that all the regressions are "correct".

I think the actual issue is just that the managed fallback is deferring to the static `GetElement` and so its forcing the entire 16-byte vector to dereference rather than directly accessing the relevant element (hence causing a nullref prior to the bounds check). Instead we should be doing the bounds check up front and then deferring to `GetElementUnsafe` to ensure the behavior is deterministic.

**Comment by tannergooding** [comment]
That is when `DOTNET_EnableAVX=0` we're doing:
```csharp
Vector256<float> tmp = this;
if (index < 0 || index >= Count)
    throw ArgumentOutOfRangeException();
return Unsafe.ReadUnaligned(ref Unsafe.As<Vector256<float>, float>(ref tmp), index);
```

While when we are accelerated we're doing:
```
if (index < 0 || index >= Count)
    throw ArgumentOutOfRangeException();
return Unsafe.ReadUnaligned(ref Unsafe.As<Vector256<float>, float>(ref tmp), index);
```

and therefore not doing the full dereference first

**Comment by jakobbotsch** [comment]
@tannergooding The JIT is internally inconsistent because of this bug -- so this cannot just be fixed on the managed side. We have `GetElement` tree that looks like this:
```scala
N008 ( 11, 13) [000007] ---XG------                         └──▌  HWINTRINSIC float  float GetElement <l:$2c3, c:$2c2>
N002 (  3,  2) [000001] ---XG------                            ├──▌  IND       simd8  <l:$240, c:$241>
N001 (  1,  1) [000000] -----------                            │  └──▌  LCL_VAR   byref  V00 this         u:1 (last use) $80
N007 (  7, 10) [000006] ---X-------                            └──▌  COMMA     int    $280
N005 (  6,  9) [000005] ---X-------                               ├──▌  BOUNDS_CHECK_ArgRng void   $203
N003 (  1,  1) [000004] -----------                               │  ├──▌  LCL_VAR   int    V01 arg1         u:1 $c0
N004 (  1,  1) [000003] -----------                               │  └──▌  CNS_INT   int    2 $44
N006 (  1,  1) [000002] -----------                               └──▌  LCL_VAR   int    V01 arg1         u:1 (last use) $c0
```
This requires the indirection to be evaluated before the bounds check. The lowering transformation being fixed here is reordering the evaluation order of the operands. Alternatively this IR should not have been created like this in the first place, but that's still a fix somewhere else within the JIT.

https://github.com/dotnet/runtime/pull/83005#issuecomment-1653663453 is the original context.

**Comment by tannergooding** [comment]
I think the shape here is wrong in that we've produced `COMMA(BOUNDS_CHECK_ArgRng, op2)`. This should've been `COMMA(BOUNDS_CHECK_ArgRng, GetElement(op1, op2))`

The bounds check is meant to prevent the intrinsic (and any operand evaluation) from executing. That is, it's meant to be the first thing done before we do anything else.

**Comment by tannergooding** [comment]
Although I guess that would still be potentially problematic with regards to non accelerated handling since `GetElement` takes `Vector256<T>` by value
and therefore it'd force the indirection to occur before get element happens

so I guess the real issue is we didn't spill op1 when inserting the bounds check during import

**Pull request merged**